File: browser/src/lib/pb/wire.ts
// Hand-written protojson mirror of shared/proto/appunvs.proto.
// On-the-wire enum values are short lowercase strings; field names are snake_case.

export const Role = {
  UNSPECIFIED: 'role_unspecified',
  PROVIDER: 'provider',
  CONNECTOR: 'connector',
  BOTH: 'both'
} as const;
export type Role = (typeof Role)[keyof typeof Role];

export const Op = {
  UNSPECIFIED: 'op_unspecified',
  UPSERT: 'upsert',
  DELETE: 'delete',
  TABLE_CREATE: 'table_create',
  TABLE_DELETE: 'table_delete',
  COLUMN_ADD: 'column_add',
  COLUMN_DELETE: 'column_delete',
  QUOTA_EXCEEDED: 'quota_exceeded'
} as const;
export type Op = (typeof Op)[keyof typeof Op];

export const Platform = {
  UNSPECIFIED: 'platform_unspecified',
  BROWSER: 'browser',
  DESKTOP: 'desktop',
  MOBILE: 'mobile'
} as const;
export type Platform = (typeof Platform)[keyof typeof Platform];

export interface Message {
  seq?: number;
  device_id: string;
  user_id: string;
  namespace: string;
  role: Role;
  op: Op;
  table: string;
  payload?: Record<string, unknown>;
  ts: number;
}

export interface RegisterRequest {
  device_id: string;
  platform: Platform;
}

export interface RegisterResponse {
  token: string;
  user_id: string;
}

function isValidRole(v: unknown): v is Role {
  return v === 'role_unspecified' || v === 'provider' || v === 'connector' || v === 'both';
}
function isValidOp(v: unknown): v is Op {
  return (
    v === 'op_unspecified' ||
    v === 'upsert' ||
    v === 'delete' ||
    v === 'table_create' ||
    v === 'table_delete' ||
    v === 'column_add' ||
    v === 'column_delete' ||
    v === 'quota_exceeded'
  );
}

/** Serialize a Message to canonical protojson. Matches the Go reference:
 *  default-valued scalars (empty strings, zero i64) are omitted. */
export function toJson(m: Message): string {
  const out: Record<string, unknown> = {};
  if (m.seq !== undefined && m.seq !== 0) out.seq = m.seq;
  if (m.device_id) out.device_id = m.device_id;
  if (m.user_id) out.user_id = m.user_id;
  if (m.namespace) out.namespace = m.namespace;
  if (m.role && m.role !== 'role_unspecified') out.role = m.role;
  if (m.op && m.op !== 'op_unspecified') out.op = m.op;
  if (m.table) out.table = m.table;
  if (m.payload !== undefined) out.payload = m.payload;
  if (m.ts !== undefined && m.ts !== 0) out.ts = m.ts;
  return JSON.stringify(out);
}

/** Parse a canonical protojson Message. Numeric fields tolerate string form
 *  (protojson renders int64 as string). */
export function fromJson(s: string): Message {
  const raw = JSON.parse(s) as Record<string, unknown>;

  const seqVal = raw.seq;
  const tsVal = raw.ts;
  const role = raw.role;
  const op = raw.op;

  if (!isValidRole(role)) throw new Error(`invalid role: ${String(role)}`);
  if (!isValidOp(op)) throw new Error(`invalid op: ${String(op)}`);

  const m: Message = {
    device_id: String(raw.device_id ?? ''),
    user_id: String(raw.user_id ?? ''),
    namespace: String(raw.namespace ?? ''),
    role,
    op,
    table: String(raw.table ?? ''),
    ts: numeric(tsVal) ?? 0
  };
  const seq = numeric(seqVal);
  if (seq !== undefined) m.seq = seq;
  if (raw.payload && typeof raw.payload === 'object') {
    m.payload = raw.payload as Record<string, unknown>;
  }
  return m;
}

function numeric(v: unknown): number | undefined {
  if (v === undefined || v === null) return undefined;
  if (typeof v === 'number') return v;
  if (typeof v === 'string' && v !== '') {
    const n = Number(v);
    return Number.isFinite(n) ? n : undefined;
  }
  return undefined;
}

File: desktop/src-tauri/src/wire.rs
//! Hand-written protojson mirror of `shared/proto/appunvs.proto`.
//!
//! Wire format: canonical protojson with `UseProtoNames` enabled.
//! - Field names are snake_case (enforced via `#[serde(rename_all = "snake_case")]`).
//! - Enum values serialize as short lowercase strings (`"provider"`, `"upsert"`, ...).
//! - `seq` is omitted when zero so clients don't accidentally claim a sequence.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    RoleUnspecified,
    Provider,
    Connector,
    Both,
}

impl Role {
    pub fn as_wire(self) -> &'static str {
        match self {
            Role::RoleUnspecified => "role_unspecified",
            Role::Provider => "provider",
            Role::Connector => "connector",
            Role::Both => "both",
        }
    }

    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "provider" => Some(Role::Provider),
            "connector" => Some(Role::Connector),
            "both" => Some(Role::Both),
            "role_unspecified" => Some(Role::RoleUnspecified),
            _ => None,
        }
    }

    pub fn includes_provider(self) -> bool {
        matches!(self, Role::Provider | Role::Both)
    }

    pub fn includes_connector(self) -> bool {
        matches!(self, Role::Connector | Role::Both)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Op {
    OpUnspecified,
    Upsert,
    Delete,
    TableCreate,
    TableDelete,
    ColumnAdd,
    ColumnDelete,
    QuotaExceeded,
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Platform {
    #[default]
    PlatformUnspecified,
    Browser,
    Desktop,
    Mobile,
}

fn is_zero(v: &i64) -> bool {
    *v == 0
}
fn is_role_unspecified(r: &Role) -> bool {
    matches!(r, Role::RoleUnspecified)
}
fn is_op_unspecified(o: &Op) -> bool {
    matches!(o, Op::OpUnspecified)
}

/// Wire envelope. Matches `appunvs.v1.Message`.
///
/// Scalar fields with a default value are omitted on the wire (proto3 /
/// protojson convention). Parsing tolerates any subset of fields for forward
/// and backward compatibility with schema broadcasts that don't carry a
/// device_id, etc.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    #[serde(default, skip_serializing_if = "is_zero")]
    pub seq: i64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub device_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub user_id: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub namespace: String,
    #[serde(default = "default_role", skip_serializing_if = "is_role_unspecified")]
    pub role: Role,
    #[serde(default = "default_op", skip_serializing_if = "is_op_unspecified")]
    pub op: Op,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub table: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<Map<String, Value>>,
    #[serde(default, skip_serializing_if = "is_zero")]
    pub ts: i64,
}

fn default_role() -> Role {
    Role::RoleUnspecified
}
fn default_op() -> Op {
    Op::OpUnspecified
}

impl Message {
    pub fn new_upsert(
        device_id: &str,
        user_id: &str,
        role: Role,
        table: &str,
        payload: Map<String, Value>,
        ts: i64,
    ) -> Self {
        Self {
            seq: 0,
            device_id: device_id.to_string(),
            user_id: user_id.to_string(),
            namespace: user_id.to_string(),
            role,
            op: Op::Upsert,
            table: table.to_string(),
            payload: Some(payload),
            ts,
        }
    }

    pub fn new_delete(
        device_id: &str,
        user_id: &str,
        role: Role,
        table: &str,
        id: &str,
        ts: i64,
    ) -> Self {
        let mut p = Map::new();
        p.insert("id".to_string(), Value::String(id.to_string()));
        Self {
            seq: 0,
            device_id: device_id.to_string(),
            user_id: user_id.to_string(),
            namespace: user_id.to_string(),
            role,
            op: Op::Delete,
            table: table.to_string(),
            payload: Some(p),
            ts,
        }
    }

    pub fn payload_id(&self) -> Option<&str> {
        self.payload.as_ref()?.get("id")?.as_str()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisterRequest {
    pub device_id: String,
    pub platform: Platform,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisterResponse {
    pub token: String,
    pub user_id: String,
}

/// Payloads for /auth/signup and /auth/login — mirror appunvs.v1.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthCredentials {
    pub email: String,
    pub password: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionResponse {
    pub user_id: String,
    pub session_token: String,
}

/// Response shape for /auth/me.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MeResponse {
    #[serde(default)]
    pub user_id: String,
    #[serde(default)]
    pub email: String,
    #[serde(default)]
    pub created_at: i64,
    #[serde(default)]
    pub devices: Vec<RemoteDevice>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RemoteDevice {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub user_id: String,
    #[serde(default)]
    pub platform: Platform,
    #[serde(default)]
    pub created_at: i64,
    #[serde(default)]
    pub last_seen: i64,
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn role_serializes_as_lowercase_short_name() {
        let j = serde_json::to_string(&Role::Provider).unwrap();
        assert_eq!(j, "\"provider\"");
    }

    #[test]
    fn message_omits_seq_when_zero() {
        let m = Message {
            seq: 0,
            device_id: "d1".into(),
            user_id: "u1".into(),
            namespace: "u1".into(),
            role: Role::Provider,
            op: Op::Upsert,
            table: "records".into(),
            payload: None,
            ts: 1,
        };
        let j = serde_json::to_string(&m).unwrap();
        assert!(!j.contains("\"seq\""), "seq should be omitted when zero: {}", j);
    }
}
